tests/descriptor_tests/test_github.py

```diff
@@ -0,0 +1,369 @@
+
+    @skip_remote_access_check(has_access=False)
+    def test_get_constraint_cached_release(self):
+        """
+        Test that the get_latest_cached_version() method correctly finds the latest
+        acceptable release on disk when a constraint pattern is provided.
+        """
+        with patch(_TESTED_CLASS + "._get_locally_cached_versions") as cached_versions_mock:
+            # Ensure that the latest cached version that matches the provided constraint pattern
+            # is returned (and not a newer one that doesn't match the constraint pattern.)
+            cached_versions_dict = {"v3.2.1": "/fake/path", "v4.2.1": "/faker/path"}
+            cached_versions_mock.return_value = cached_versions_dict
+            desc = self._create_desc()
+            desc2 = desc.find_latest_cached_version(constraint_pattern="v3.x.x")
+            self.assertEqual(desc2.get_version(), "v3.2.1")
+            # If no cached versions match the provided constraint pattern, None should be returned.
+            desc2 = desc.find_latest_cached_version(constraint_pattern="v5.x.x")
+            self.assertEqual(desc2, None)
+
+    def test_has_remote_access(self):
+        """
+        Test that the has_remote_access() method returns as expected when able and unable to
+        connect to the Github API, and requests the correct URL.
+        """
+        desc = self._create_desc()
+        target_url = "https://api.github.com/repos/{o}/{r}"
+        target_url = target_url.format(
+            o=self.default_location_dict["organization"],
+            r=self.default_location_dict["repository"]
+        )
+        with patch(_TESTED_MODULE + ".urllib2.urlopen") as urlopen_mock:
+            # normal good response
+            urlopen_mock.return_value = MockResponse("repo_root")
+            self.assertEqual(desc.has_remote_access(), True)
+            urlopen_mock.assert_called_with(target_url)
+            # 404 response
+            urlopen_mock.side_effect = MockResponse("repo_root_404").get_exception()
+            self.assertEqual(desc.has_remote_access(), False)
+            # No internet connection, no response from GH API, etc
+            urlopen_mock.side_effect = urllib2.URLError("Test exception.")
+            self.assertEqual(desc.has_remote_access(), False)
+
+    @skip_remote_access_check()
+    def test_bundle_cache_path(self):
+        """
+        Test that the bund_cache_path is built as expected.
+        """
+        desc = self._create_desc()
+        expected_bundle_cache_path = os.path.join(
+            self.bundle_cache,
+            "github",
+            self.default_location_dict["organization"],
+            self.default_location_dict["repository"],
+            self.default_location_dict["version"],
+        )
+        found_bundle_cache_path = desc._io_descriptor._get_bundle_cache_path(self.bundle_cache)
+        self.assertEqual(found_bundle_cache_path, expected_bundle_cache_path)
+
+    @skip_remote_access_check()
```

Comment:
this may be a stupid question, but could we split things up so that the remote access check happens separately - then we could avoid all the decorators?